src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp

```diff
@@ -677,95 +732,199 @@ bool load_environment_server_info(
                         {
                             // A server cannot be reach in all interfaces, it's clearly a localhost call
-                            IPLocator::setIPv4(server_locator, "127.0.0.1");
+                            IPLocator::setIPv6(server_locator, "::1");
+                        }
+
+                        // get port if any
+                        int port = DEFAULT_ROS2_SERVER_PORT;
+                        if (++it != mr.cend() && it->matched)
+                        {
+                            port = stoi(it->str());
+                        }
+
+                        process_port( port, server_locator);
+                    }
+
+                    // add server to the list
+                    add_server2qos(server_id, std::forward_list<Locator>{server_locator}, attributes);
+                }
+                // try resolve DNS
+                else if (std::regex_match(locator, mr, ROS2_DNS_DOMAINPORT_PATTERN,
+                        std::regex_constants::match_not_null))
+                {
+                    std::forward_list<Locator> flist;
+
+                    {
+                        std::stringstream new_locator(locator,
+                                std::ios_base::in |
+                                std::ios_base::out |
+                                std::ios_base::ate);
+
+                        // first try the formal notation, add default port if necessary
+                        if (!mr[2].matched)
+                        {
+                            new_locator << ":" << DEFAULT_ROS2_SERVER_PORT;
                         }
 
-                        if (++it != mr.cend())
+                        new_locator >> server_locator;
+                    }
+
+                    // Otherwise add all resolved locators
+                    switch ( server_locator.kind )
+                    {
+                        case LOCATOR_KIND_UDPv4:
+                        case LOCATOR_KIND_UDPv6:
+                            flist.push_front(server_locator);
+                            break;
+                        case LOCATOR_KIND_INVALID:
                         {
-                            // reset the locator to default
-                            IPLocator::setPhysicalPort(server_locator, DEFAULT_ROS2_SERVER_PORT);
+                            std::smatch::iterator it = mr.cbegin();
 
-                            if (it->matched)
+                            // traverse submatches
+                            if (++it != mr.cend())
                             {
-                                // note stoi throws also an invalid_argument
-                                int port = stoi(it->str());
+                                std::string domain_name = it->str();
+                                std::set<std::string> ipv4, ipv6;
+                                std::tie(ipv4, ipv6) = IPLocator::resolveNameDNS(domain_name);
```

Comment:
Yes because is not the same.
+ The `new_locator` *stringstream* is populated with the actual environment variable text between `;`. A  `:11811` is appended and we try to read the locator using a specific method from `Locator.h`:
```
inline std::istream& operator >>( std::istream& input, Locator_t& loc)
```
+ Only if the above method fails do we try to resolve the domain name detected using the regex `ROS2_DNS_DOMAINPORT_PATTERN`. This fallback is for the case there is no specific transport prefix *UDPvxxx*. And will add all locators available.
To summarize, if a prefix is available there is only a `IPLocator::resolveNameDNS` call. And if is not too. No redundancy.
I could have done everything via regex but if we have builtin functionality we better use it (this way any builtin improvement is automatically applied to this code).